bin/main.rb

```diff
@@ -1,76 +1,97 @@
+  [6, 7, 8], # bottom row
+  [0, 4, 8], # left to right diagonal
+  [2, 4, 6], # right to left diagonal
+  [0, 3, 6], # left-most column
+  [1, 4, 7], # middle column
+  [2, 5, 8] # right-most column
+].freeze
+
+board = %w[1 2 3 4 5 6 7 8 9]
+game_tokens = %w[X O]
+game = Game.new
+
+puts 'Welcome to tic tac toe game'
+
+# register the players
+puts 'Player 1, input name: '
+player_one_name = gets.chomp
 
-puts 'WELCOME TO TIC-TAC-TOE GAME !!!'
+puts 'Player 2, input name: '
+player_two_name = gets.chomp
 
-# create an empty game board
-@board = %w[1 2 3 4 5 6 7 8 9]
+player_one = game.create_player(player_one_name, nil, 0)
+player_two = game.create_player(player_two_name, nil, 0)
 
-# create game tokens as an array of 2 items
-@game_tokens = %w[X O]
+# allow players to chose their tokens
+game.select_tokens(player_one, player_two, game_tokens)
 
-# create an empty hash to hold player and token info for each player
-@assigned_tokens = {}
-
-# register two players
-puts 'Player 1, input name: '
-@player_one = gets.chomp
-
-puts 'Player 2, input name: '
-@player_two = gets.chomp
-
-# allow first player choose a game token X or O
-puts "#{@player_one}, Choose a token: 'X' or 'O'"
-selected_token = gets.chomp.upcase
-
-# validate the token chosen
-if !@game_tokens.to_s.include? selected_token
-  puts "Invalid token, please select 'X' or 'O'"
-else
-  # assign the selected_token to player_one and the other token to player_two
-  puts 'Tokens assigned successfully'
-  puts "#{@player_one}'s Token:  #{@assigned_tokens[:player_one]}"
-  puts "#{@player_two}'s Token:  #{@assigned_tokens[:player_two]}"
-end
+puts 'Let the game begin !!!'
+# take players' inputs for game play
 
 no_winner = false
 winner_found = false
-
-# determine first player
-
-# call current player method to determine who to play first
+last_token = nil
+current_player = nil
 
 while !winner_found || !no_winner
-  # display gameboard
+
   # prompt current player to select a move which is an index
+  current_token = game.switch_turn(player_one, player_two, last_token)
+
+  if current_token == player_one.token
+    current_player = player_one
+    puts "#{player_one.name}'s Turn (#{current_token})"
+  else
+    current_player = player_two
+    puts "#{player_two.name}'s Turn (#{current_token})"
+  end
+
+  # display current board before player's entry
+  game.display_board(board)
+
   puts 'Enter a position: '
   position = gets.chomp
-  puts position
 
-  # validate the position. If already taken or if out of available board index,
-  # display message and prompt for the input again.
+  # translate player's input into board index
+  board_index = game.input_to_index(position.to_i, board)
 
-  # if position is valid,
-  # update board with the current player's token
+  # update board with the current player's token at chosen board index
+  board_position = board_index.pred
+  board[board_position] = current_token
+  last_token = current_token
+  current_player.move_count += 1
```

Comment:
Move it into the `Game` class wrapped into a method called for example `update_board`